File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Paging.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;

using Raven.Abstractions.Data;
using Raven.Client;
using Raven.Client.Document;
using Raven.Client.Indexes;
using Raven.Documentation.CodeSamples.Orders;

namespace Raven.Documentation.Samples.Indexes.Querying
{
	public class Paging
	{
		#region paging_0_4
		public class Products_ByUnitsInStock : AbstractIndexCreationTask<Product>
		{
			public Products_ByUnitsInStock()
			{
				Map = products => from product in products
								  select new
								  {
									  UnitsInStock = product.UnitsInStock
								  };
			}
		}
		#endregion

		#region paging_6_0
		public class Orders_ByOrderLines_ProductName : AbstractIndexCreationTask<Order>
		{
			public Orders_ByOrderLines_ProductName()
			{
				Map = orders => from order in orders
								from line in order.Lines
								select new
								{
									Product = line.ProductName
								};

				MaxIndexOutputsPerDocument = 1024;
			}
		}
		#endregion

		public Paging()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_0_1
					IList<Product> results = session
						.Query<Product, Products_ByUnitsInStock>()
						.Where(x => x.UnitsInStock > 10)
						.ToList();
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_0_2
					IList<Product> results = session
						.Advanced
						.DocumentQuery<Product, Products_ByUnitsInStock>()
						.WhereGreaterThan(x => x.UnitsInStock, 10)
						.ToList();
					#endregion
				}

				#region paging_0_3
				QueryResult result = store
					.DatabaseCommands
					.Query(
						"Products/ByUnitsInStock",
						new IndexQuery
						{
							Query = "UnitsInStock_Range:{Ix10 TO NULL}"
						});
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_1_1
					IList<Product> results = session
						.Query<Product, Products_ByUnitsInStock>()
						.Where(x => x.UnitsInStock > 10)
						.Take(9999)	// server will decrease this value to 1024
						.ToList();
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_1_2
					IList<Product> results = session
						.Advanced
						.DocumentQuery<Product, Products_ByUnitsInStock>()
						.WhereGreaterThan(x => x.UnitsInStock, 10)
						.Take(9999)	// server will decrease this value to 1024
						.ToList();
					#endregion
				}

				#region paging_1_3
				QueryResult result = store
					.DatabaseCommands
					.Query(
						"Products/ByUnitsInStock",
						new IndexQuery
						{
							Query = "UnitsInStock_Range:{Ix10 TO NULL}",
							PageSize = 9999
						});
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_2_1
					IList<Product> results = session
						.Query<Product, Products_ByUnitsInStock>()
						.Where(x => x.UnitsInStock > 10)
						.Skip(20)	// skip 2 pages worth of products
						.Take(10)	// take up to 10 products
						.ToList();	// execute query
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_2_2
					IList<Product> results = session
						.Advanced
						.DocumentQuery<Product, Products_ByUnitsInStock>()
						.WhereGreaterThan(x => x.UnitsInStock, 10)
						.Skip(20)	// skip 2 pages worth of products
						.Take(10)	// take up to 10 products
						.ToList();	// execute query
					#endregion
				}

				#region paging_2_3
				QueryResult result = store
					.DatabaseCommands
					.Query(
						"Products/ByUnitsInStock",
						new IndexQuery
						{
							Query = "UnitsInStock_Range:{Ix10 TO NULL}",
							Start = 20,	// skip 2 pages worth of products
							PageSize = 10	// take up to 10 products
						});
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_3_1
					RavenQueryStatistics stats;
					IList<Product> results = session
						.Query<Product, Products_ByUnitsInStock>()
						.Statistics(out stats)				// fill query statistics
						.Where(x => x.UnitsInStock > 10)
						.Skip(20)
						.Take(10)
						.ToList();

					var totalResults = stats.TotalResults;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_3_2
					RavenQueryStatistics stats;
					IList<Product> results = session
						.Advanced
						.DocumentQuery<Product, Products_ByUnitsInStock>()
						.Statistics(out stats)					// fill query statistics
						.WhereGreaterThan(x => x.UnitsInStock, 10)
						.Skip(20)
						.Take(10)
						.ToList();

					var totalResults = stats.TotalResults;
					#endregion
				}
			}

			using (var store = new DocumentStore())
			{
				#region paging_3_3
				QueryResult result = store
					.DatabaseCommands
					.Query(
						"Products/ByUnitsInStock",
						new IndexQuery
						{
							Query = "UnitsInStock_Range:{Ix10 TO NULL}",
							Start = 20,
							PageSize = 10
						});

				var totalResults = result.TotalResults;
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_4_1
					RavenQueryStatistics stats;
					IList<Product> results;
					var pageNumber = 0;
					var pageSize = 10;
					var skippedResults = 0;

					do
					{
						results = session
							.Query<Product, Products_ByUnitsInStock>()
							.Statistics(out stats)
							.Skip((pageNumber * pageSize) + skippedResults)
							.Take(pageSize)
							.Where(x => x.UnitsInStock > 10)
							.Distinct()
							.ToList();

						skippedResults += stats.SkippedResults;
						pageNumber++;
					}
					while (results.Count > 0);
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_4_2
					RavenQueryStatistics stats;
					IList<Product> results;
					var pageNumber = 0;
					var pageSize = 10;
					var skippedResults = 0;

					do
					{
						results = session
							.Advanced
							.DocumentQuery<Product, Products_ByUnitsInStock>()
							.Statistics(out stats)
							.Skip((pageNumber * pageSize) + skippedResults)
							.Take(pageSize)
							.WhereGreaterThan(x => x.UnitsInStock, 10)
							.Distinct()
							.ToList();

						skippedResults += stats.SkippedResults;
						pageNumber++;
					}
					while (results.Count > 0);
					#endregion
				}
			}

			using (var store = new DocumentStore())
			{
				#region paging_4_3
				QueryResult result;
				var pageNumber = 0;
				var pageSize = 10;
				var skippedResults = 0;

				do
				{
					result = store
						.DatabaseCommands
						.Query(
							"Products/ByUnitsInStock",
							new IndexQuery
							{
								Query = "UnitsInStock_Range:{Ix10 TO NULL}",
								Start = (pageNumber * pageSize) + skippedResults,
								PageSize = pageSize,
								IsDistinct = true
							});

					skippedResults += result.SkippedResults;
					pageNumber++;
				}
				while (result.Results.Count > 0);
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_6_1
					RavenQueryStatistics stats;
					IList<Order> results;
					var pageNumber = 0;
					var pageSize = 10;
					var skippedResults = 0;

					do
					{
						results = session
							.Query<Order, Orders_ByOrderLines_ProductName>()
							.Statistics(out stats)
							.Skip((pageNumber * pageSize) + skippedResults)
							.Take(pageSize)
							.ToList();

						skippedResults += stats.SkippedResults;
						pageNumber++;
					}
					while (results.Count > 0);
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region paging_6_2
					RavenQueryStatistics stats;
					IList<Order> results;
					var pageNumber = 0;
					var pageSize = 10;
					var skippedResults = 0;

					do
					{
						results = session
							.Advanced
							.DocumentQuery<Order, Orders_ByOrderLines_ProductName>()
							.Statistics(out stats)
							.Skip((pageNumber * pageSize) + skippedResults)
							.Take(pageSize)
							.ToList();

						skippedResults += stats.SkippedResults;
						pageNumber++;
					}
					while (results.Count > 0);
					#endregion
				}
			}

			using (var store = new DocumentStore())
			{
				#region paging_6_3
				QueryResult result;
				var pageNumber = 0;
				var pageSize = 10;
				var skippedResults = 0;

				do
				{
					result = store
						.DatabaseCommands
						.Query(
							"Orders/ByOrderLines/ProductName",
							new IndexQuery
							{
								Start = (pageNumber * pageSize) + skippedResults,
								PageSize = pageSize
							});

					skippedResults += result.SkippedResults;
					pageNumber++;
				}
				while (result.Results.Count > 0);
				#endregion
			}

			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region paging_5_1
					var pagingInformation = new RavenPagingInformation();
					IList<Product> results = session
						.Advanced
						.LoadStartingWith<Product>(
							"products/",				// all documents starting with 'products/'
							"1*|2*",				// rest of the key must begin with "1" or "2" e.g. products/10, products/25
							0 * 25,					// skip 0 records (page 1)
							25,					// take up to 25
							pagingInformation: pagingInformation);	// fill `RavenPagingInformation` with operation data

					results = session
						.Advanced
						.LoadStartingWith<Product>(
							"products/",				// all documents starting with 'products/'
							"1*|2*",				// rest of the key must begin with "1" or "2" e.g. products/10, products/25
							1 * 25,					// skip 25 records (page 2)
							25,					// take up to 25
							pagingInformation: pagingInformation);	// since this is a next page to 'page 1' and we are passing 'RavenPagingInformation' that was filled during 'page 1' retrieval, rapid pagination will take place
					#endregion
				}
			}
		}
	}
}

File: Documentation/3.5/Samples/java/src/test/java/net/ravendb/indexes/querying/Paging.java
package net.ravendb.indexes.querying;

import java.util.List;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.data.IndexQuery;
import net.ravendb.abstractions.data.QueryResult;
import net.ravendb.abstractions.indexing.FieldStorage;
import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.RavenPagingInformation;
import net.ravendb.client.RavenQueryStatistics;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.indexes.AbstractIndexCreationTask;
import net.ravendb.samples.northwind.Order;
import net.ravendb.samples.northwind.Product;
import net.ravendb.samples.northwind.QProduct;


public class Paging {

  //region paging_0_4
  public static class Products_ByUnitsInStock extends AbstractIndexCreationTask {
    public Products_ByUnitsInStock() {
      map =
       " from product in docs.Products            " +
       " select new                               " +
       " {                                        " +
       "     UnitsInStock = product.UnitsInStock  " +
       " };";
    }
  }
  //endregion

  //region paging_6_0
  public static class Orders_ByOrderLines_ProductName extends AbstractIndexCreationTask {
    @SuppressWarnings("boxing")
    public Orders_ByOrderLines_ProductName() {
      map =
       " from order in docs.Orders " +
       " from line in order.Lines       " +
       " select new                     " +
       " {                              " +
       "     Product = line.ProductName " +
       " };";

       maxIndexOutputsPerDocument = 1024L;
    }
  }
  //endregion

  //region paging_7_0
  public static class Orders_ByStoredProductName extends AbstractIndexCreationTask {
    public static class Result {
      private String product;
      public String getProduct() {
        return product;
      }
      public void setProduct(String product) {
        this.product = product;
      }
    }
    @SuppressWarnings("boxing")
    public Orders_ByStoredProductName() {
      map =
       " from order in docs.Orders " +
       " from line in order.Lines       " +
       " select new                     " +
       " {                              " +
       "     Product = line.ProductName " +
       " };";

       maxIndexOutputsPerDocument = 1024L;
       store("Product", FieldStorage.YES);
    }
  }
  //endregion

  @SuppressWarnings({"unused", "boxing"})
  public Paging() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_0_1
        QProduct p = QProduct.product;
        List<Product> results = session
          .query(Product.class, Products_ByUnitsInStock.class)
          .where(p.unitsInStock.gt(10))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_0_2
        QProduct p = QProduct.product;
        List<Product> results = session
          .advanced()
          .documentQuery(Product.class, Products_ByUnitsInStock.class)
          .whereGreaterThan(p.unitsInStock, 10)
          .toList();
        //endregion
      }

      //region paging_0_3
      QueryResult result = store
        .getDatabaseCommands()
        .query("Products/ByUnitsInStock",
          new IndexQuery("UnitsInStock_Range:{Ix10 TO NULL}"));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_1_1
        QProduct p = QProduct.product;
        List<Product> results = session
          .query(Product.class, Products_ByUnitsInStock.class)
          .where(p.unitsInStock.gt(10))
          .take(9999) // server will decrease this value to 1024
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_1_2
        QProduct p = QProduct.product;
        List<Product> results = session
          .advanced()
          .documentQuery(Product.class, Products_ByUnitsInStock.class)
          .whereGreaterThan(p.unitsInStock, 10)
          .take(9999) // server will decrease this value to 1024
          .toList();
        //endregion
      }

      //region paging_1_3
      IndexQuery query = new IndexQuery();
      query.setQuery("UnitsInStock_Range:{Ix10 TO NULL}");
      query.setPageSize(9999);
      QueryResult result = store
        .getDatabaseCommands()
        .query("Products/ByUnitsInStock", query);
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_2_1
        QProduct p = QProduct.product;
        List<Product> results = session
          .query(Product.class, Products_ByUnitsInStock.class)
          .where(p.unitsInStock.gt(10))
          .skip(20) // skip 2 pages worth of products
          .take(10) // take up to 10 products
          .toList(); // execute query
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_2_2
        QProduct p = QProduct.product;
        List<Product> results = session
          .advanced()
          .documentQuery(Product.class, Products_ByUnitsInStock.class)
          .whereGreaterThan(p.unitsInStock, 10)
          .skip(20) // skip 2 pages worth of products
          .take(10) // take up to 10 products
          .toList(); // execute query
        //endregion
      }

      //region paging_2_3
      IndexQuery query = new IndexQuery();
      query.setQuery("UnitsInStock_Range:{Ix10 TO NULL}");
      query.setStart(20); // skip 2 pages worth of products
      query.setPageSize(10); // take up to 10 products

      QueryResult result = store
        .getDatabaseCommands()
        .query("Products/ByUnitsInStock", query);
      //endregion
    }


    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_3_1
        QProduct p = QProduct.product;
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Product> results = session
          .query(Product.class, Products_ByUnitsInStock.class)
          .statistics(statsRef) // fill query statistics
          .where(p.unitsInStock.gt(10))
          .skip(20)
          .take(10)
          .toList();

        int totalResults = statsRef.value.getTotalResults();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_3_2
        QProduct p = QProduct.product;
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Product> results = session
          .advanced()
          .documentQuery(Product.class, Products_ByUnitsInStock.class)
          .statistics(statsRef) // fill query statistics
          .whereGreaterThan(p.unitsInStock, 10)
          .skip(20)
          .take(10)
          .toList();

        int totalResults = statsRef.value.getTotalResults();
        //endregion
      }
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region paging_3_3
      IndexQuery query = new IndexQuery();
      query.setQuery("UnitsInStock_Range:{Ix10 TO NULL}");
      query.setStart(20);
      query.setPageSize(10);
      QueryResult result = store
        .getDatabaseCommands()
        .query("Products/ByUnitsInStock", query);

      int totalResults = result.getTotalResults();
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_4_1
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Product> results;
        int pageNumber = 0;
        int pageSize = 10;
        int skippedResults = 0;

        QProduct p = QProduct.product;

        do {
          results = session
            .query(Product.class, Products_ByUnitsInStock.class)
            .statistics(statsRef)
            .skip((pageNumber * pageSize) + skippedResults)
            .take(pageSize)
            .where(p.unitsInStock.gt(10))
            .distinct()
            .toList();

          skippedResults += statsRef.value.getSkippedResults();
          pageNumber++;
        } while (results.size() > 0);
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_4_2
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Product> results;
        int pageNumber = 0;
        int pageSize = 10;
        int skippedResults = 0;

        QProduct p = QProduct.product;

        do {
          results = session
            .advanced()
            .documentQuery(Product.class, Products_ByUnitsInStock.class)
            .statistics(statsRef)
            .skip((pageNumber * pageSize) + skippedResults)
            .take(pageSize)
            .whereGreaterThan(p.unitsInStock, 10)
            .distinct()
            .toList();

          skippedResults += statsRef.value.getSkippedResults();
          pageNumber++;
        } while (results.size() > 0);
        //endregion
      }
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region paging_4_3
      QueryResult result;
      int pageNumber = 0;
      int pageSize = 10;
      int skippedResults = 0;

      do {
        IndexQuery query = new IndexQuery();
        query.setQuery("UnitsInStock_Range:{Ix10 TO NULL}");
        query.setStart((pageNumber * pageSize) + skippedResults);
        query.setPageSize(pageSize);
        query.setDistinct(true);
        result = store
          .getDatabaseCommands()
          .query("Products/ByUnitsInStock", query);

        skippedResults += result.getSkippedResults();
        pageNumber++;
      } while (result.getResults().size() > 0);

      //endregion
    }


    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_6_1
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Order> results;
        int pageNumber = 0;
        int pageSize = 10;
        int skippedResults = 0;

        do {
          results = session
            .query(Order.class, Orders_ByOrderLines_ProductName.class)
            .statistics(statsRef)
            .skip((pageNumber * pageSize) + skippedResults)
            .take(pageSize)
            .toList();

          skippedResults += statsRef.value.getSkippedResults();
          pageNumber++;
        } while  (results.size() > 0);
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_6_2
        Reference<RavenQueryStatistics> statsRef = new Reference<RavenQueryStatistics>();
        List<Order> results;
        int pageNumber = 0;
        int pageSize = 10;
        int skippedResults = 0;

        do {
          results = session
            .advanced()
            .documentQuery(Order.class, Orders_ByOrderLines_ProductName.class)
            .statistics(statsRef)
            .skip((pageNumber * pageSize) + skippedResults)
            .take(pageSize)
            .toList();

          skippedResults += statsRef.value.getSkippedResults();
          pageNumber++;
        } while  (results.size() > 0);
        //endregion
      }
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region paging_6_3
      QueryResult result;
      int pageNumber = 0;
      int pageSize = 10;
      int skippedResults = 0;

      do {
        IndexQuery query = new IndexQuery();
        query.setStart((pageNumber * pageSize) + skippedResults);
        query.setPageSize(pageSize);

        result = store
          .getDatabaseCommands()
          .query("Orders/ByOrderLines/ProductName", query);

        skippedResults += result.getSkippedResults();
        pageNumber++;
      } while (result.getResults().size() > 0);
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_5_1
        RavenPagingInformation pagingInformation = new RavenPagingInformation();
        Product[] results = session
          .advanced()
          .loadStartingWith(Product.class,
            "products/", // all documents starting with 'products/'
            "1*|2*", // rest of the key must begin with "1" or "2" e.g. products/10, products/25
            0 * 25,  // skip 0 records (page 1)
            25,  // take up to 25
            null,
            pagingInformation); // fill `RavenPagingInformation` with operation data

        results = session
          .advanced()
          .loadStartingWith(Product.class,
            "products/", // fill `RavenPagingInformation` with operation data
            "1*|2*",  // rest of the key must begin with "1" or "2" e.g. products/10, products/25
            1 * 25, // skip 25 records (page 2)
            25, // take up to 25
            null,
            pagingInformation); // since this is a next page to 'page 1' and we are passing 'RavenPagingInformation' that was filled during 'page 1' retrieval, rapid pagination will take place
        //endregion
      }
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region paging_7_1
        List<String> results;
        int pageNumber = 0;
        int pageSize = 10;

        do {
          results = session
             .query(Orders_ByStoredProductName.Result.class, Orders_ByStoredProductName.class)
             .select(String.class, "Product")
             .skip(pageSize * pageNumber)
             .take(pageSize)
             .distinct()
             .toList();

          pageNumber++;
        } while (results.size() > 0);
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region paging_7_2
        List<Orders_ByStoredProductName.Result> results;
        int pageNumber = 0;
        int pageSize = 10;

        do {
          results = session
             .advanced()
             .documentQuery(Order.class, Orders_ByStoredProductName.class)
             .selectFields(Orders_ByStoredProductName.Result.class, "Product")
             .skip(pageSize * pageNumber)
             .take(pageSize)
             .distinct()
             .toList();

          pageNumber++;
        } while (results.size() > 0);
        //endregion
      }

      //region paging_7_3
      QueryResult result;
      int pageNumber = 0;
      int pageSize = 10;

      do {
        IndexQuery query = new IndexQuery();
        query.setStart(pageNumber * pageSize);
        query.setPageSize(pageSize);
        query.setDistinct(true);
        query.setFieldsToFetch(new String[] { "product" });

        result = store.getDatabaseCommands()
           .query("Orders/ByStoredProductName", query);
        pageNumber++;
      } while (result.getResults().size() > 0);
      //endregion
    }
  }
}
